src/modes/ModeEstimation.jl

```diff
@@ -0,0 +1,205 @@
+using ..Turing
+using ..Bijectors
+using LinearAlgebra
+
+import ..DynamicPPL
+import ..DynamicPPL: Model, AbstractContext, VarInfo
+import Optim
+import Optim: optimize
+import NamedArrays
+import ..ForwardDiff
+import StatsBase
+import Printf
+
+export MAP, MLE, optimize
+
+struct MLE end
+struct MAP end
+
+"""
+    OptimLogDensity{M<:Model,C<:Context,V<:VarInfo}
+
+A struct that stores the log density function of a `DynamicPPL` model.
+"""
+struct OptimLogDensity{M<:Model,C<:AbstractContext,V<:VarInfo}
+    "A `DynamicPPL.Model` constructed either with the `@model` macro or manually."
+    model::M
+    "A `DynamicPPL.AbstractContext` used to evaluate the model. `LikelihoodContext` or `DefaultContext` are typical for MAP/MLE."
+    context::C
+    "A `DynamicPPL.VarInfo` struct that will be used to update model parameters."
+	init::V
+end
+
+"""
+    OptimLogDensity(model::Model, context::AbstractContext)
+
+Create a callable `OptimLogDensity` struct that evaluates a model using the given `context`.
+"""
+function OptimLogDensity(model::Model, context::AbstractContext)
+	init = VarInfo(model)
+	DynamicPPL.link!(init, SampleFromPrior())
+	return OptimLogDensity(model, context, init)
+end
+
+"""
+    (f::OptimLogDensity)(z; unlinked::Bool = false)
+
+Evaluate the log joint (with `DefaultContext`) or log likelihood (with `LikelihoodContext`)
+at the array `z`. If `unlinked=true`, no change of variables will occur.
+"""
+function (f::OptimLogDensity)(z; unlinked::Bool = false)
+    spl = DynamicPPL.SampleFromPrior()
+
+    varinfo = DynamicPPL.VarInfo(f.init, spl, z)
+
+    unlinked && DynamicPPL.invlink!(f.init, spl)
```

Comment:
Do I understand it correctly that this is done in each iteration. Is it really necessary to check this at each iteration, or could you pull out the code instead and do this once?